src/replication.c

```diff
@@ -3528,269 +3757,114 @@ void syncWithPrimary(connection *conn) {
+            return;
         }
-        sdsfree(err);
         server.repl_state = REPL_STATE_RECEIVE_IP_REPLY;
         return;
-    }
-
-    if (server.repl_state == REPL_STATE_RECEIVE_IP_REPLY && !server.replica_announce_ip)
-        server.repl_state = REPL_STATE_RECEIVE_CAPA_REPLY;
-
     /* Receive REPLCONF ip-address reply. */
-    if (server.repl_state == REPL_STATE_RECEIVE_IP_REPLY) {
-        err = receiveSynchronousResponse(conn);
-        if (err == NULL) goto no_response_error;
-        /* Ignore the error if any, not all the Redis OSS versions support
-         * REPLCONF ip-address. */
-        if (err[0] == '-') {
-            serverLog(LL_NOTICE,
-                      "(Non critical) Primary does not understand "
-                      "REPLCONF ip-address: %s",
-                      err);
+    case REPL_STATE_RECEIVE_IP_REPLY:
+        if (syncWithPrimaryHandleReceiveIPReplyState(conn) == C_ERR) {
+            syncWithPrimaryHandleError(&conn);
+            return;
         }
-        sdsfree(err);
         server.repl_state = REPL_STATE_RECEIVE_CAPA_REPLY;
-        return;
-    }
-
+        if (server.replica_announce_ip) return;
+        /* fall through */
     /* Receive CAPA reply. */
-    if (server.repl_state == REPL_STATE_RECEIVE_CAPA_REPLY) {
-        err = receiveSynchronousResponse(conn);
-        if (err == NULL) goto no_response_error;
-        /* Ignore the error if any, not all the Redis OSS versions support
-         * REPLCONF capa. */
-        if (err[0] == '-') {
-            serverLog(LL_NOTICE,
-                      "(Non critical) Primary does not understand "
-                      "REPLCONF capa: %s",
-                      err);
+    case REPL_STATE_RECEIVE_CAPA_REPLY:
+        if (syncWithPrimaryHandleReceiveCapaReplyState(conn) == C_ERR) {
+            syncWithPrimaryHandleError(&conn);
+            return;
         }
-        sdsfree(err);
-        err = NULL;
         server.repl_state = REPL_STATE_RECEIVE_VERSION_REPLY;
         return;
-    }
-
     /* Receive VERSION reply. */
-    if (server.repl_state == REPL_STATE_RECEIVE_VERSION_REPLY) {
-        err = receiveSynchronousResponse(conn);
-        if (err == NULL) goto no_response_error;
-        /* Ignore the error if any. Valkey >= 8 supports REPLCONF VERSION. */
-        if (err[0] == '-') {
-            serverLog(LL_NOTICE,
-                      "(Non critical) Primary does not understand "
-                      "REPLCONF VERSION: %s",
-                      err);
+    case REPL_STATE_RECEIVE_VERSION_REPLY:
+        if (syncWithPrimaryHandleReceiveVersionReplyState(conn) == C_ERR) {
+            syncWithPrimaryHandleError(&conn);
+            return;
         }
-        sdsfree(err);
-        err = NULL;
         if (server.cluster_enabled) {
             server.repl_state = REPL_STATE_RECEIVE_NODEID_REPLY;
             return;
         } else {
             server.repl_state = REPL_STATE_SEND_PSYNC;
+            goto case_send_psync;
```

Comment:
Not the cleanest solution, but could not think of another way to do this

Reply:
that's ok. let's merge now and address this later.